src/net/udp.rs

```diff
@@ -1,26 +1,92 @@
-use bytes::{Buf, BufMut, Bytes};
+use bytes::Bytes;
 use tokio::sync::{mpsc, Mutex};
 
 use crate::{
     envelope::{Datagram, Protocol},
     ToSocketAddrs, World, TRACING_TARGET,
 };
 
-use std::{cmp, io::Result, net::SocketAddr};
+use std::{
+    cmp,
+    io::{self, Result},
+    net::SocketAddr,
+};
 
 /// A simulated UDP socket.
 ///
 /// All methods must be called from a host within a Turmoil simulation.
 pub struct UdpSocket {
     local_addr: SocketAddr,
-    rx: Mutex<mpsc::Receiver<(Datagram, SocketAddr)>>,
+    rx: Mutex<Rx>,
+}
+
+struct Rx {
```

Comment:
Nice solution! I'm going to take on a task to refactor this into the host similar to what TCP is doing. That should simplify some things.